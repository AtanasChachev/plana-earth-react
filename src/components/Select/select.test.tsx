import { screen, render, fireEvent, act, within } from '@testing-library/react';
import { waitForElementToBeRemoved } from '@testing-library/dom';
import SelectComponent from './Select';

const MOCK_OPTIONS = [
  {
    id: 'test',
    name: 'Test name',
  }, 
  {
    id: 'test_two',
    name: 'Test two',
  },
];

const onChange = jest.fn();
const setup = () => {
  const { asFragment } = render(
    <SelectComponent 
      options={MOCK_OPTIONS}
      placeholder='Test Select Component'
      onChange={onChange}
    />,
  );

  const select = screen.getByTestId('select');

  return { asFragment, select };
};

const openSelectDropdown = () => {
  const { select } = setup();
  const selectButton = within(select).getByRole('button');

  act(() => {
    fireEvent.mouseDown(selectButton);
  });

  const selectDropdown = document.querySelector('.MuiPopover-root') as HTMLElement;

  return { select, selectDropdown };
};

const triggerDropdownOption = () => {
  const { select, selectDropdown } = openSelectDropdown();
  const selectDropdownOption = selectDropdown.querySelectorAll('.MuiPopover-root [role="option"]')[0] as HTMLElement;

  act(() => {
    fireEvent.click(selectDropdownOption);
  });

  return { select };
};

describe('<Select />', () => {
  test('<Select /> should render properly', () => {
    const { asFragment } = setup();

    expect(asFragment()).toMatchSnapshot();
  });

  test('<Select /> should open popup', () => {
    const { selectDropdown } = openSelectDropdown();
    expect(selectDropdown).toBeInTheDocument();
  });

  test('<Select /> dropdown click should execute onChange', () => {
    triggerDropdownOption();
    expect(onChange).toHaveBeenCalledTimes(1);
  });

  test('<Select /> dropdown option click should umount the component from DOM', async () => {
    triggerDropdownOption();

    const selectContainerAfterManipulation = await waitForElementToBeRemoved(() =>
      document.querySelector('.MuiPopover-root'),
    );

    expect(selectContainerAfterManipulation).toBeUndefined();
  });

  test('<Select /> dropdown option click should update input text value', () => {
    const { select } = triggerDropdownOption();
    const selectInputTextWrapper = within(select).getByRole('button');

    expect(selectInputTextWrapper.textContent).toEqual('Test name');
  });
});