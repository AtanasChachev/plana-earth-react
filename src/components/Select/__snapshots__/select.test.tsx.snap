// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> <Select /> should render properly 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root select "
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root select__label MuiInputLabel-formControl MuiInputLabel-animated"
      data-shrink="false"
    >
      Test Select Component
    </label>
    <div
      aria-required="true"
      class="MuiInputBase-root MuiInput-root MuiInput-underline select__form MuiInputBase-formControl MuiInput-formControl"
      data-testid="select"
    >
      <div
        aria-haspopup="listbox"
        class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
        role="button"
        tabindex="0"
      >
        <span>
          â€‹
        </span>
      </div>
      <input
        aria-hidden="true"
        class="MuiSelect-nativeInput"
        tabindex="-1"
        value=""
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSelect-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
